muon/_prot/preproc.py

```diff
@@ -193,30 +195,52 @@ def dsb(
     return toreturn
 
 
-def clr(adata: AnnData, inplace: bool = True) -> Union[None, AnnData]:
+def clr(adata: AnnData, inplace: bool = True, axis: int = 0) -> Union[None, AnnData]:
     """
     Apply the centered log ratio (CLR) transformation
     to normalize counts in adata.X.
 
     Args:
         data: AnnData object with protein expression counts.
         inplace: Whether to update adata.X inplace.
+        axis: Axis across which CLR is performed.
     """
+
+    if not inplace:
+        adata = adata.copy()
+
     sparse = False
     if issparse(adata.X):
         sparse = True
     # Geometric mean of ADT counts
     x = adata.X
-    g_mean = np.exp(np.log1p(x).sum(axis=0) / x.shape[0])
-    # Centered log ratio
-    clr = np.log1p(x / g_mean)
 
-    if sparse:
-        clr = csr_matrix(clr)
+    if axis == 1:
+        if issparse(x):
+            if not isinstance(x, csr_matrix):
+                logging.warning(
+                    "`adata.X` is not stored in CSR format. To perform CLR along `axis=1`, either "
+                    "convert `adata.X` to CSR format or densify it."
+                )
+                return
```

Comment:
I went for a simple warning and not performing CLR in order not stick with the inplace solution. Not sure if this is desirable, TBH. If we convert to a CSR format internally, this approach where the data is not densified, could also be applied for `axis=0` using CSC formal, i.e. s.t. along the lines of

```python
if axis == 1 and issparse(adata.X) and not isinstance(adata.X, csr_matrix):
    x = csr_matrix(adata.X)
elif axis == 0 and issparse(adata.X) and not isinstance(adata.X, csc_matrix):
    x = csc_matrix(adata.X)
else:
    x = adata.X

if issparse(x):
    x.data /= np.repeat(
        np.exp(np.log1p(x).sum(axis=axis).A ...
    )
    np.log1p(x.data, out=x.data)
else:
    np.log1p(...)

# convert back to original sparse format
if sparse:
    ...

adata.X = x

return None if inplace else adata
```

Would be more readable, IMO. What do you think @gtca, @ilia-kats?

Reply:
Yes, I think this makes sense. But I don't think that we should convert back to the original sparsity format afterwards. I would just throw a warning and return the converted CSC/CSR matrix, something like
```python
x = adata.X
if issparse(adata.X):
    if axis == 1 and not isinstance(x, csr_matrix):
        warnings.warn("adata.X is sparse but not in CSR format. Converting to CSR")
        x = csr_matrix(x)
    elif axis == 0 and not isinstance(x, csc_matrix):
        warnings.warn("adata.X is sparse but not in CSC format. Converting to CSC")
        x = csc_matrix(x)
    x.data = ...
else:
    x = ...

adata.X = x
```